File: BaekJoon/차이를 최대로.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <stack>

using namespace std;

/*
차이를 최대로.
입력받은 수들의 차이의 절대값들의 합이 최대가 되게끔 하는 최대값을 출력.

->정렬하고 순열로 다 돌려서 각각 계산하고 max로 출력해도 정답이 나올만큼 입력수가 작음.
*/
int n;

int calculate(int* a) {
	int res = 0;
	for (int i = 1; i < n; i++) {
		res += abs(a[i-1]-a[i]);
	}
	return res;
}
bool next_permutation(int* a, int n) {
	int i = n - 1;
	while (i > 0 && a[i - 1] > a[i]) i -= 1;
	if (i <= 0) return false;
	int j = n - 1;
	//while (a[i - 1] <= a[j]) j -= 1;
	while (a[i - 1] >= a[j]) j -= 1;
	swap(a[i - 1], a[j]);
	j = n - 1;
	while (i < j) {
		swap(a[i], a[j]);
		i += 1;
		j -= 1;
	}
	return true;
}
int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);

	n = 0;
	cin >> n;

	int* a = new int[n];
	for (int i = 0; i < n; i++) {
		cin >> a[i];
	}

	sort(a, a + n);

	int ans = 0;
	bool t = next_permutation(a, n);
	do {
		int temp = calculate(a);
		ans = max(ans, temp);
		t = next_permutation(a, n);
	} while (t);

	cout << ans << endl;

}

File: BaekJoon/차이를 최대로.java
import java.util.*;

/*
순열 비교할때 while문의 값비교가 =도 포함되어야 순열에서 같은값에 대한 처리가 가능하다. 아니면 while문을 못벗어나서
시간초과가 뜸.
 */
public class 차이를최대로 {
    //public static int n;
    //public static int[] a;
    //public static int ans=0;

    public static boolean next_permutation(int[] a){
        int i=a.length-1;
        while(i>0 && a[i-1]>=a[i]){
            i-=1;
        }
        if(i<=0) return false;
        int j=a.length-1;
        while(j>=i && a[i-1]>=a[j]){
            j-=1;
        }
        int temp=a[i-1];
        a[i-1]=a[j];
        a[j]=temp;

        j=a.length-1;
        while(i<j){
            temp=a[i];
            a[i]=a[j];
            a[j]=temp;
            i++;
            j--;
        }
        return true;
    }
    public static int calc(int[] a){
        int sum=0;
        for(int i=0;i<a.length-1;i++) {
            sum+=Math.abs(a[i]-a[i+1]);
        }
        return sum;
    }

    public static void main(String[] args) {
        Scanner sc=new Scanner(System.in);
        int n;
        int[] a;
        n=sc.nextInt();
        a=new int[n];
        for(int i=0;i<n;i++){
            a[i]=sc.nextInt();
        }
        Arrays.sort(a);
        int ans=0;
        do{
            int temp=calc(a);
            ans=Math.max(ans,temp);
        }while(next_permutation(a));

        System.out.println(ans);
    }
}
